src/crypto/ecdh.rs

```diff
@@ -249,7 +360,71 @@ pub fn encrypt<R: CryptoRng + Rng>(
 
     let encrypted_key_len = vec![u8::try_from(encrypted_key.len())?];
 
-    Ok(vec![encoded_public, encrypted_key_len, encrypted_key])
+    Ok(EncryptionOutput {
+        encoded_public,
+        encrypted_key_len,
+        encrypted_key,
+    })
+}
+
+fn encrypt_p521<R: CryptoRng + Rng>(
+    rng: &mut R,
+    alg_sym: SymmetricKeyAlgorithm,
+    hash: HashAlgorithm,
+    q: &[u8],
+    plain: &[u8],
+    param: &[u8],
+) -> Result<EncryptionOutput> {
+    use p521::{ecdh::EphemeralSecret, EncodedPoint, PublicKey};
+
+    let their_public = PublicKey::from_sec1_bytes(q)?;
+
+    let our_secret = EphemeralSecret::random(rng);
+
+    // derive shared secret
+    let shared_secret = our_secret.diffie_hellman(&their_public);
+
+    let z = kdf(
+        hash,
+        shared_secret.raw_secret_bytes().as_ref(),
+        alg_sym.key_size(),
+        param,
+    )?;
+
+    let mut plain_padded = plain.to_vec();
+    let plain_padded_ref = pad(&mut plain_padded)?;
+
+    // Perform AES Key Wrap
+    let encrypted_key = aes_kw::wrap(&z, plain_padded_ref)?;
+
+    // Public point gets encoded with prefix with 0x04
+    let encoded_public = EncodedPoint::from(PublicKey::from(&our_secret))
+        .to_bytes()
+        .to_vec();
+
+    let encrypted_key_len = vec![u8::try_from(encrypted_key.len())?];
+
+    Ok(EncryptionOutput {
+        encoded_public,
+        encrypted_key_len,
+        encrypted_key,
+    })
+}
+
+/// PKCS5 pads the plaintext and returns the block-aligned slice of the padded plaintext
+fn pad(plain: &mut Vec<u8>) -> Result<&[u8]> {
+    // PKCS5 padding (PKCS5 is PKCS7 with a blocksize of 8)
+    let len = plain.len();
+    plain.resize(len + BLOCK_SIZE, 0);
+
+    let pos = len;
+    let bs = BLOCK_SIZE * (pos / BLOCK_SIZE);
+    if plain.len() < bs || plain.len() - bs < BLOCK_SIZE {
+        bail!("unable to pad");
+    }
+    let buf = GenericArray::<u8, U8>::from_mut_slice(&mut plain[bs..bs + BLOCK_SIZE]);
```

Comment:
Can avoid needing to add to the offset with the starting index by slicing twice

```suggestion
    let buf = GenericArray::<u8, U8>::from_mut_slice(&mut plain[bs..][..BLOCK_SIZE]);
```